sdk/program/src/loader_upgradeable_instruction.rs

```diff
@@ -0,0 +1,104 @@
+
+#[repr(u8)]
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Clone)]
+pub enum UpgradeableLoaderInstruction {
+    /// Initialize a Buffer account.
+    ///
+    /// A Buffer account is an intermediary that once fully populated is used
+    /// with the `DeployWithMaxDataLen` instruction to populate the program's
+    /// ProgramData account.
+    ///
+    /// The `InitializeBuffer` instruction requires no signers and MUST be
+    /// included within the same Transaction as the system program's
+    /// `CreateAccount` instruction that creates the account being initialized.
+    /// Otherwise another party may initialize the account.
+    ///
+    /// # Account references
+    ///   0. [writable] source account to initialize.
+    InitializeBuffer,
+
+    /// Write program data into a Buffer account.
+    ///
+    /// # Account references
+    ///   0. [writable, signer] Buffer account to write program data to.
+    Write {
+        /// Offset at which to write the given bytes.
+        offset: u32,
+        /// Serialized program data
+        #[serde(with = "serde_bytes")]
+        bytes: Vec<u8>,
+    },
+
+    /// Deploy an executable program.
+    ///
+    /// A program consists of a Program and ProgramData account pair.
+    ///   - The Program account's address will serve as the program id any
+    ///     instructions that execute this program.
+    ///   - The ProgramData account will remain mutable by the loader only and
+    ///     holds the program data and authority information.  The ProgramData
+    ///     account's address is derived from the Program account's address and
+    ///     created by the DeployWithMaxDataLen instruction.
+    ///
+    /// The ProgramData address is derived from the Program account's address as
+    /// follows:
+    ///
+    /// `let (program_data_address, _) = Pubkey::find_program_address(
+    ///      &[program_address],
+    ///      &bpf_loader_upgradeable::id()
+    ///  );`
+    ///
+    /// The `DeployWithMaxDataLen` instruction does not require the ProgramData
+    /// account be a signer and therefore MUST be included within the same
+    /// Transaction as the system program's `CreateAccount` instruction that
+    /// creates the Program account. Otherwise another party may initialize
+    /// the account.
+    ///
+    /// # Account references
+    ///   0. [Signer] The payer account that will pay to create the ProgramData
+    ///      account.
+    ///   1. [writable] The uninitialized ProgramData account.
+    ///   2. [writable] The uninitialized Program account.
+    ///   3. [writable] The Buffer account where the program data has been
+    ///      written.
+    ///   4. [] Rent sysvar.
+    ///   5. [] Clock sysvar.
+    ///   6. [] System program (`solana_sdk::system_program::id()`).
+    ///   7. [] The program's authority, optional, if omitted then the program
+    ///      will not be upgradeable.
+    DeployWithMaxDataLen {
```

Comment:
Just call this `Deploy`?

Reply:
Thought we were saving that for later when realloc was supported

Reply:
When relloc is supported, I guess the max data len here essentially becomes ignored right?

Reply:
Correct

Reply:
I'm unclear about what you're thinking we'd change when realloc is supported.   A new `Deploy` instruction?  Or just keep `DeployWithMaxDataLen` but now `max_data_len` is just totally ignored?